### ISSUE TITLE
add extension.importMode={exclusive, inclusive} and update import feed to follow this rule

### ISSUE THREAD

**Issue by elbidone** [meumobi/sitebuilder#108]
RSS extension should be configured to define if importation is exclusive or inclusive (this option also exists for csv import). 
- Exclusive: affect all items. The db should contains exactly items within RSS.
- Inclusive: affect only items within RSS. New items on RSS are added to db, but items saved on a previous import and missing on latest remains valid

**Comment by juliogreff** [comment]
We'll create a service **BulkImportItems**. Every other service that wants to bulk-import items will use this one. For example, CSV import.

```php
class BulkImportItems {
    public function perform($category, $items, $mode, $shouldCreate, $shouldUpdate) {
        $shouldCreate = $shouldCreate ?: function($existing, new) { return true };
        $shouldUpdate = $shouldUpdate ?: function($existing, new) {
            return $new->id && in_array($new, $existing);
        };
        $existingItems = $category->items;
        $importedItems = [];

        foreach ($items as $item) {
            if ($shouldUpdate($existingItems, $item)) {
                $importedItems []= $this->updateItem($item);
            } else if ($shouldCreate($existingItems, $item)) {
                $importedItems []= $this->createItem($item);
            }

            $importedItems []= $item;
        }

        if ($mode == EXCLUSIVE) {
            $itemsToDelete = array_diff($existingItems, $importedItems);
            foreach ($itemsToDelete as $item) $item->delete();
        }
    }
}
```

This service alone is not enough to implement the feature though, since this was designed to be very generic. That's why the shouldCreate and shouldUpdate parameters are for (they might need better names though, feel free to suggest). They're callbacks that decide if the item should be created, updated, or not acted upon. If they're not acted upon, we assume they're just left alone, so we add them to the importedItems collection (**note**: maybe rename that collection to *itemsToKeep* instead?). I don't know how to check for the identity of the item in that case, but let's see some code before we decide that.

From UpdateNewsFeed, those two parameters should do:

```
// this is WIP, I'll come back and edit later
$shouldUpdate = $existing->find([ 'guid' => $item->guid, 'checksum' => ['$ne' => $item->checksum] ])
```

**Comment by t4deu** [comment]
The BulkImportItems has updated to follow the recent changes made in the UpdateNewsFeed refactory.

To make the BulkImportItems a generic service used , it shoud not implement the "createItem" or "updateItem", Instead I'm using the methods of the import service, passed as a parameter.

```php
class BulkImportItems {
    public function perform($params) {
        $importedItemsIds = [];
    	  list($extension, $category, $importService, $mode) =  ParamsValidator::validate($params, [...]);
    		
        $feed = $importService->fetchFeed($extension);
        $items = $importService->extractItems($feed)
				
        foreach ($items as $item) {
          if ($importService->shouldUpdate($category, $item)) {
              $importedItemsIds []= $importService->updateItem($item)->id;
          } else if ($importService->shouldCreate($category, $item)) {
              $importedItemsIds []= $importService->createItem($item)->id;
          }
      	}

        if ($mode == EXCLUSIVE) {
        	Items::remove(['_id' => ['$nin' => $ids]]);                                        
        }
    }
}
```

Rename UpdateEventsService to UpdateEventsFeed and implement the necessary methods, following the current implementation of UpdateNewsFeed, this service will be passed as parameter to the BulkImportItems:
```php
class UpdateEventsService
{
	public function shouldCreate($category, $item)
	{
	  return true
	}
	
	public function shouldUpdate($category, $item)
	{
	  return !Events::find('count', [                                               
      'conditions' => [                                                          
        'parent_id' => $category->id,                                            
        'guid' => $item['guid'],                                              
      ],                                                                         
    ]); 
	}
	
	public function updateItem($category, $item)
	{
	  $event = Events::find('type', ['conditions' => [                                                                                           
      '_id' => $id                                                                                                                                     
    ]]);
    return $event->set($item)->save();
	}
	
	public function createItem($category, $item)
	{
	  return Events::create($item)->save();
	}
	
	public function fetchFeed($extension)
	{
	  //TODO
	}
	
	public function extraxtItems($feed)
	{
	  //TODO
	}
}
```

**Comment by juliogreff** [comment]
Bulk import should know nothing about feeds. You can think that CSV import will also use this service. It does have createItem and updateItem, which are simple calls to the create and update services.

What you pass as parameters are callbacks, not classes. Please review your analysis.

**Comment by t4deu** [comment]
Include the "import_mode" field in the Extension model:
```php
//app/models/Extensions.php
public static function __init()                                                
{                                                                                                                        
  $self->_schema = $parent->_schema + array(                                   
    'import_mode' => array('type' => 'integer', 'default' => BulkImportItems::InclusiveMode),
  );
```

Add the import mode to the extension form:
```php
//app/views/extensions/_form.htm.php 
<div class="form-grid-570">                                                  
	<?= $this->form->input('import_method', array(                            
		'type' => 'radio',                                                   
		'value' => BulkImportItems::InclusiveMode,                                                        
		'options' => array(s('Inclusive'), s('Exclusive'))                   
	)) ?>                                                                        
</div> 
```

**Comment by t4deu** [comment]
Can't use the BulkImportItems service in the UpdateNewsFeed because it not allows create an item passing the ItemCreation options. Not allowing send push notification jobs.

Any help to solve is welcome, but pass the services or more callbacks as BulkImportItems parameters isn't an option, as recomended by @juliogreff

But I'm looking for the best way to solve it

**Comment by t4deu** [comment]
Pass the CreateService options as parameter to send push notifications when importing items:
```php
  public function perform($params)                                                  
  {                                                                                 
    list($category, $items, $mode, $shouldCreate, $shouldUpdate, $sendPush) =  ParamsValidator::validate($params, [
      'category',                                                                   
      'items',                                                                      
      'mode',                                                                    
      'shouldCreate',                                                            
      'shouldUpdate',
      'sendPush',                                                            
    ]);                                                                          
                                                                                                                                                                  
    foreach ($items as $item) {                                                  
      if ($shouldUpdate($item) && $this->updateItem($item)) {                    
        $importedItemsIds[] = $item->id();                                       
       } else if ($shouldCreate($item) && $this->createItem($item, $sendPush)) {             
        $importedItemsIds[] = $item->id();                                       
       }                                                                          
    }  
```

**Comment by juliogreff** [comment]
Closed in #187
